File: backend/models/message.py
import enum

from sqlalchemy import (
    Column,
    DateTime,
    Enum,
    ForeignKey,
    Integer,
    String,
    Text,
)
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from database import Base
from schemas.callback import MessageType


class MessageFrom:
    CUSTOMER = 1
    USER = 2
    LLM = 3


class MessageStatus:
    """Business logic status (kept for backward compatibility)"""
    PENDING = 1
    REPLIED = 2
    RESOLVED = 3
    ESCALATED = 4  # Farmer requested human help after AI response


class DeliveryStatus(enum.Enum):
    """Twilio delivery status tracking"""
    PENDING = "PENDING"          # Initial state, not sent yet
    QUEUED = "QUEUED"            # Twilio accepted, queuing
    SENDING = "SENDING"          # Twilio sending to WhatsApp
    SENT = "SENT"                # Sent to WhatsApp servers
    DELIVERED = "DELIVERED"      # Delivered to device
    READ = "READ"                # Customer read (if callbacks enabled)
    FAILED = "FAILED"            # Permanent failure
    UNDELIVERED = "UNDELIVERED"  # Temporary failure/expired


class MediaType(enum.Enum):
    """Type of media in message"""
    TEXT = "TEXT"              # Regular text message (default)
    VOICE = "VOICE"            # Voice/audio message
    IMAGE = "IMAGE"            # Image (future)
    VIDEO = "VIDEO"            # Video (future)
    DOCUMENT = "DOCUMENT"      # PDF/document (future)
    LOCATION = "LOCATION"      # Geolocation (future)
    OTHER = "OTHER"            # Unknown media type


class Message(Base):
    __tablename__ = "messages"

    id = Column(Integer, primary_key=True, index=True)
    message_sid = Column(String, unique=True, index=True, nullable=False)
    customer_id = Column(Integer, ForeignKey("customers.id"), nullable=False)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    body = Column(Text, nullable=False)
    from_source = Column(Integer, nullable=False)
    message_type = Column(Enum(MessageType), nullable=True)

    # Business logic status (kept for backward compatibility)
    status = Column(
        Integer,
        nullable=False,
        server_default=str(MessageStatus.PENDING),
        index=True,
    )

    # Delivery tracking fields
    delivery_status = Column(
        Enum(DeliveryStatus),
        nullable=False,
        server_default=DeliveryStatus.PENDING.value,
        index=True,
    )
    twilio_error_code = Column(String(10), nullable=True)
    twilio_error_message = Column(Text, nullable=True)
    retry_count = Column(Integer, nullable=False, server_default="0")
    last_retry_at = Column(DateTime(timezone=True), nullable=True)
    delivered_at = Column(DateTime(timezone=True), nullable=True)

    # Media tracking (for voice, images, etc.)
    media_url = Column(String, nullable=True)
    media_type = Column(
        Enum(MediaType),
        nullable=False,
        server_default=MediaType.TEXT.value,
    )

    created_at = Column(DateTime(timezone=True), server_default=func.now())

    customer = relationship("Customer", back_populates="messages")
    user = relationship("User")

    def is_delivery_failed(self) -> bool:
        """Check if message delivery permanently failed"""
        return self.delivery_status in (
            DeliveryStatus.FAILED, DeliveryStatus.UNDELIVERED
        )

    def can_retry(self, max_retries: int = 3) -> bool:
        """Check if message can be retried"""
        return (
            self.delivery_status in (
                DeliveryStatus.PENDING, DeliveryStatus.FAILED
            )
            and self.retry_count < max_retries
        )

File: app/database/dao/types/message.ts
// Message types and interfaces
// Note: from_source uses MessageFrom constants (1=CUSTOMER, 2=USER, 3=LLM)
// Note: status uses MessageStatus constants (1=PENDING, 2=REPLIED, 3=RESOLVED)
// Note: message_type (1=REPLY, 2=WHISPER)
// Note: is_used (0=not used, 1=used) - tracks if WHISPER message has been accepted
// Note: delivery_status - Twilio delivery tracking: PENDING, QUEUED, SENDING, SENT, DELIVERED, READ, FAILED, UNDELIVERED
// Note: media_type - TEXT, VOICE, IMAGE, VIDEO, DOCUMENT, LOCATION, OTHER
// Import from @/constants/messageSource and @/constants/messageStatus when needed
export interface Message {
  id: number;
  from_source: number; // 1=CUSTOMER, 2=USER, 3=LLM
  message_sid: string;
  customer_id: number;
  user_id: number | null;
  body: string;
  message_type: number;
  status: number; // 1=PENDING, 2=REPLIED, 3=RESOLVED
  is_used: number; // 0=not used, 1=used (for WHISPER messages)
  delivery_status: string; // Twilio delivery status: PENDING, QUEUED, SENDING, SENT, DELIVERED, READ, FAILED, UNDELIVERED
  media_url: string | null; // URL to media file (images, voice, etc.)
  media_type: string; // TEXT, VOICE, IMAGE, VIDEO, DOCUMENT, LOCATION, OTHER
  createdAt: string;
  updatedAt: string;
}

export interface CreateMessageData {
  from_source: number; // Use MessageFrom constants
  message_sid: string;
  customer_id: number;
  user_id: number | null;
  body: string;
  createdAt: string;
  message_type?: number | string;
  status?: number; // Use MessageStatus constants (default: PENDING=1)
  is_used?: number; // 0=not used, 1=used (default: 0)
  delivery_status?: string; // Twilio delivery status (default: PENDING)
  media_url?: string | null; // URL to media file (images, voice, etc.)
  media_type?: string; // TEXT, VOICE, IMAGE, etc. (default: TEXT)
}

export interface UpdateMessageData {
  from_source?: number;
  message_sid?: string;
  customer_id?: number;
  user_id?: number | null;
  body?: string;
  message_type?: number | string;
  status?: number; // Use MessageStatus constants
  is_used?: number; // 0=not used, 1=used
  delivery_status?: string; // Twilio delivery status
  media_url?: string | null; // URL to media file (images, voice, etc.)
  media_type?: string; // TEXT, VOICE, IMAGE, etc.
}

// Extended message interface with user details (for inbox/conversation views)
// Note: Inherits media_url and media_type from Message interface
export interface MessageWithUsers extends Message {
  customer_name: string;
  customer_phone: string;
  user_name: string | null; // Full name of user who sent the message (for USER messages)
  user_email: string | null; // Email of user who sent the message (for USER messages)
}

// Conversation summary for inbox
export interface ConversationSummary {
  customer_id: number;
  user_id: number;
  customer_name: string;
  customer_phone: string;
  user_name: string;
  user_email: string;
  last_message: string;
  last_message_type: string;
  last_message_time: string;
  unread_count: number;
}
